gleambatch.py

```diff
@@ -0,0 +1,227 @@
+            gv2.set_beta(mit["param_beta"])
+            gv2id = "{}.574".format(random.randint(1700000000000, 1800000000000))
+            gv2.set_id(gv2id)
+            gv2.set_name(gv2.full_name(batch.name))
+            sim = Simulation(gv2, None)
+            batch.add_simulation_info(
+                sim,
+                name=sce["name"],
+                group=mit["label"],
+                color=sce.get("color"),
+                style=sce.get("style"),
+            )
+
+
+def generate(args):
+    "Run the 'generate' subcommand"
+
+    with open(args.CONFIG_YAML, "rt") as f:
+        config = yaml.load(f)
+    Path(config["output_dir"]).expanduser().mkdir(exist_ok=True, parents=True)
+    batch = Batch.new(config, suffix=args.comment.replace(" ", "-"))
+
+    log.info(f"Reading regions from {batch.config['regions_file']} ...")
+    rs = Regions.load_from_yaml(batch.config["regions_file"])
+
+    estimate(batch, rs)
+
+    gv = estimates_to_gleamdef(batch, rs, args.GLEAM_XML)
+
+    parameterize(batch, gv)
+
+    batch.save_sim_defs_to_gleam()
+
+    batch.save()
+
+    log.info(
+        f"Run '{sys.argv[0]} process {batch.get_batch_file_path()}' after"
+        " running and retrieving simulations in GleamViz (and closing Gleamviz)."
+    )
+
+
+def process(args):
+    "Run the 'process' subcommand"
+
+    batch = Batch.load(args.BATCH_YAML)
+    log.info(f"Reading regions from {batch.config['regions_file']} ...")
+    rs = Regions.load_from_yaml(batch.config["regions_file"])
+    batch.load_sims(allow_unfinished=args.allow_missing)
+    batch.write_country_plots(rs)
+
+
+def create_parser():
+    ap = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    ap.add_argument(
+        "-d", "--debug", action="store_true", help="Display debugging mesages."
+    )
+    sp = ap.add_subparsers(title="subcommands", required=True, dest="cmd")
+
+    updatep = sp.add_parser("update", help="Fetch/update data from CSSE and Foretold")
+    updatep.add_argument("CONFIG_YAML", help="YAML config to use.")
+    updatep.set_defaults(func=update_data)
+
+    genp = sp.add_parser("generate", help="Create a new batch, generate GLEAM configs")
+    genp.add_argument("CONFIG_YAML", help="YAML config to use.")
+    genp.add_argument("GLEAM_XML", help="Use given XML as GLEAM def template.")
+    genp.add_argument("-c", "--comment", default="", help="Optional name comment.")
+    genp.set_defaults(func=generate)
+
+    procp = sp.add_parser(
+        "process", help="Process finished simulations from a batch, generate graphs.",
+    )
+    procp.add_argument("BATCH_YAML", help="Batch config to use.")
+    procp.set_defaults(func=process)
+    procp.add_argument(
+        "-M",
+        "--allow_missing",
+        action="store_true",
+        help="Allow missing simulation results.",
+    )
+
+    return ap
+
+
+def main():
+    logging.basicConfig(level=logging.INFO)
+    args = create_parser().parse_args()
```

Comment:
IMO should be simply `parse_args()` function which would create this... No need for additional call.

Reply:
:man_shrugging: 

Philosophical: The conceptual barrier is better here: the returned parser could be tested, serialized (theoretically), customized or used to parse different string. Your option sets the boundary around behavior rather than concepts.

Reply:
My option is not taking that way - if that's your goal, you can still have it by adding:
```
def parse_args():
  return create_parser().parse_args()
```
and keeping the main function as simple as possible (which was my intention - to reduce the reader's load). I admit it's a big detail now. I wouldn't do it before you do at least one of those fully theoretical things (up until now) as it sounds like you are building things "just in case" :grin: ... 

I have never seen serialization of a parser :grin:, but I am strawmaning on that one.